File: src/200NumberOfIslands/200NumberOfIslands.cpp
class union_find{
    vector<int> id;
    vector<int> sz;
    int cnt;
public:
    union_find(int N){
        cnt = N;
        id.resize(N);
        sz.resize(N);
        for(int i=0; i<N; ++i){
            id[i]=i;
            sz[i]=1;
        }
    }
    
    int find(int i){
        while(i!=id[i]){
            id[i]=id[id[i]];
            i=id[i];
        }
        return i;
    }
    bool connected(int p, int q){
        return find(p)==find(q);
    }
    void unioning(int p, int q){
        int i = find(p);
        int j = find(q);
        if(i==j) return;
        else if (sz[i]<sz[j]) {
            id[i]=j; sz[j]+=sz[i];
        }else{
            id[j]=i; sz[i]+=sz[j];
        }
        --cnt;
    }
    int count(){
        return cnt;
    }
};

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size(); if(m==0) return 0;    
        int n = grid[0].size(); if(n==0) return 0;
        union_find uf(m*n);
        int zeros = 0;
        for(int j=0; j<m; ++j){
            for(int i=0; i<n; ++i){
                if(grid[j][i]=='1'){
                    if(i>0&&grid[j][i-1]=='1'){
                        uf.unioning(j*n+i, j*n+i-1);
                    }    
                    if(j>0&&grid[j-1][i]=='1'){
                        uf.unioning(j*n+i, (j-1)*n+i);
                    }
                }else{
                    ++zeros;
                }
            }
        }
        return uf.count()-zeros;
    }
};

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  void dfs(vector<vector<char>> &grid, vector<vector<bool>> &visited, int i,
           int j, int m, int n) {
    visited[j][i] = true;
    int offset_x[4] = {-1, +1, 0, 0};
    int offset_y[4] = {0, 0, -1, +1};
    for (int k = 0; k < 4; k++) {
      int x = i + offset_x[k];
      int y = j + offset_y[k];
      // cout << x << " " << y << endl;
      if (x >= 0 && x < m && y >= 0 && y < n && grid[y][x] == '1' &&
          !visited[y][x]) {
        dfs(grid, visited, x, y, m, n);
      }
    }
  }
  int numIslands(vector<vector<char>> &grid) {
    int n = grid.size();
    if (n == 0) {
      return 0;
    }
    int m = grid[0].size();
    if (m == 0) {
      return 0;
    }
    vector<vector<bool>> visited;
    for (int j = 0; j < n; j++) {
      vector<bool> v(m, false);
      visited.push_back(v);
    }
    // cout << visited.size() << " " << visited[0].size() << endl;
    int num = 0;
    for (int j = 0; j < n; j++) {
      for (int i = 0; i < m; i++) {
        // cout << i << " " << j << endl;
        if (grid[j][i] == '1' && !visited[j][i]) {
          num++;
          dfs(grid, visited, i, j, m, n);
        }
      }
    }
    return num;
  }
};

int main() {
  vector<vector<char>> grid = {
      {'1', '1', '1', '1', '0'},
      {'1', '1', '0', '1', '0'},
      {'1', '1', '0', '0', '0'},
      {'0', '0', '0', '0', '0'},
  };
  Solution s;
  cout << s.numIslands(grid) << endl;
}

File: src/200NumberOfIslands/200NumberOfIslands.rs
use std::collections::VecDeque;
impl Solution {
    pub fn num_islands(grid: Vec<Vec<char>>) -> i32 {
        let m = grid.len(); if m==0 {return 0};
        let n = grid[0].len(); if n==0 {return 0};
        let mut grid = grid;
        let offset_x = vec![-1, 1,  0, 0];
        let offset_y = vec![ 0, 0, -1, 1];
        let mut queue: VecDeque<(i32, i32)> = VecDeque::new();
        let mut count = 0;
        for j in 0..m {
            for i in 0..n {
                if grid[j][i] == '1' {
                    count += 1;
                    grid[j as usize][i as usize] = '0';
                    queue.push_back((i as i32, j as i32));
                    while !queue.is_empty() {
                        let (u, v) = queue.pop_front().unwrap();                        
                        for d in 0..4 {
                            let y = v + offset_y[d];
                            let x = u + offset_x[d];
                            if y>=0 && x>=0 && y<m as i32 && x<n as i32 
                              && grid[y as usize][x as usize] == '1' {
                                grid[y as usize][x as usize] = '0';
                                queue.push_back((x as i32, y as i32));
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}
